### PR TITLE
nix: make `xcrun` visible in Nix sandbox for precompiling Metal shaders

### LINKED ISSUE #6117: Cannot find xcrun when building with Nix flakes on Darwin/MacOS
On latest version of llama.cpp, building from flake (ie `nix build github:ggerganov/llama.cpp`) produces error with latest commit (https://github.com/ggerganov/llama.cpp/commit/d01b3c4c32357567f3531d4e6ceffc5d23e87583 at time of writing). 

```
error: builder for '/nix/store/9qrdgaz8s1wfz9fwci01j2vi97y45cfy-llama-cpp-metalkit-0.0.0.drv' failed with exit code 1;
       last 25 log lines:
       > -- Generating done (0.0s)
       > CMake Warning:
       >   Manually-specified variables were not used by the project:
       >
       >     CMAKE_EXPORT_NO_PACKAGE_REGISTRY
       >     CMAKE_POLICY_DEFAULT_CMP0025
       >
       >
       > -- Build files have been written to: /tmp/nix-build-llama-cpp-metalkit-0.0.0.drv-0/source/build
       > cmake: enabled parallel building
       > cmake: enabled parallel installing
       > Running phase: buildPhase
       > build flags: -j8
       > [1/132] Compiling Metal kernels
       > FAILED: bin/default.metallib /tmp/nix-build-llama-cpp-metalkit-0.0.0.drv-0/source/build/bin/default.metallib
       > cd /tmp/nix-build-llama-cpp-metalkit-0.0.0.drv-0/source/build && xcrun -sdk macosx metal -O3 -c /tmp/nix-build-llama-cpp-metalkit-0.0.0.drv-0/source/build/bin/ggml-metal.metal -o /tmp/nix-build-llama-cpp-metalkit-0.0.0.drv-0/source/build/bin/ggml-metal.air && xcrun -sdk macosx metallib /tmp/nix-build-llama-cpp-metalkit-0.0.0.drv-0/source/build/bin/ggml-metal.air -o /tmp/nix-build-llama-cpp-metalkit-0.0.0.drv-0/source/build/bin/default.metallib && rm -f /tmp/nix-build-llama-cpp-metalkit-0.0.0.drv-0/source/build/bin/ggml-metal.air && rm -f /tmp/nix-build-llama-cpp-metalkit-0.0.0.drv-0/source/build/bin/ggml-common.h && rm -f /tmp/nix-build-llama-cpp-metalkit-0.0.0.drv-0/source/build/bin/ggml-metal.metal
       > /bin/sh: xcrun: command not found
       > [2/132] Building C object CMakeFiles/ggml.dir/ggml-alloc.c.o
       > [3/132] Building C object CMakeFiles/ggml.dir/ggml-backend.c.o
       > [4/132] Building C object CMakeFiles/ggml.dir/ggml-metal.m.o
       > [5/132] Building C object CMakeFiles/ggml.dir/ggml-quants.c.o
       > [6/132] Building C object CMakeFiles/ggml.dir/ggml.c.o
       > [7/132] Building CXX object CMakeFiles/llama.dir/unicode.cpp.o
       > [8/132] Building CXX object CMakeFiles/llama.dir/llama.cpp.o
       > ninja: build stopped: subcommand failed.
       For full logs, run 'nix log /nix/store/9qrdgaz8s1wfz9fwci01j2vi97y45cfy-llama-cpp-metalkit-0.0.0.drv'.
```

The key error is `/bin/sh: xcrun: command not found` -- Nix provides a clean `$PATH` to CMake which does not include `/usr/bin` and so Cmake is unable to locate or run `/usr/bin/xcrun`. This behavior is specific to the combination of MacOS and Nix. 

This can be fixed by either manually adding `/usr/bin` to `$PATH` prior to configuration & build steps, or by using absolute path (`/usr/bin/xcrun`) and setting `dontFixCmake = true;` in package.nix. 

System details:
Nix 2.21 (sandbox = off), MacOS 14.4 Sonoma

### PR DESCRIPTION
**Pull request by josephst**
Fixes https://github.com/ggerganov/llama.cpp/issues/6117

Adds `/usr/bin/` to $PATH prior to running Cmake when building with Nix. No change to building normally/ without Nix.

### CODE REVIEW

File: .devops/nix/package.nix
Comment (SomeoneSerge) [inline .devops/nix/package.nix:163]: I'm not particularly familiar with Darwin and I don't know what xcrun is. Could you attach a brief comment expounding this and explaining why it's needed for the `configurePhase`? What else is exposed in `/usr/bin` with the relaxed sandbox? Why do you give `/usr/bin` a higher priority than to the nix-provided `PATH`? If `xcrun` is the only thing you need from `/usr/bin` you could instead make a wrapper for it:

```
nativeBuildInputs = optionals hostPlatform.isDarwin [
  (writeShellScriptBin "xcrun" "/usr/bin/scrun")
];
```

A more general darwin question: can we explicitly request the relaxed sandbox using some kind of `requiredSystemFeatures`?

Comment (swdunlop) [reply .devops/nix/package.nix:163]: It should not be necessary to taint the build with `/usr/bin/xcrun` if all we need is `xcrun` -- it can be found in the `xcbuild` and `xcodebuild` packages.

Comment (josephst) [reply .devops/nix/package.nix:163]: @swdunlop not having any progress with `xcbuild` package -- I wonder if it's too old to know about the `metal` shader compiler. Looks like its Github project has been archived since mid-2021. 

When attempting to run `nix run nixpkgs#xcbuild` and `nix run nixpkgs#xcodebuild`, get the following error:
```
warning: non-build action metal not implemented
error: unknown build action 'metal
```

Comment (josephst) [reply .devops/nix/package.nix:163]: @SomeoneSerge `xcrun` is a MacOS/Xcode tool that finds other command line tools--for example, finding the `metal` shader compiler. I think it's used instead of a hard-coded path because different installations of Xcode can coexist, so `xcrun` will run whichever tool has been made the default by `xcode-select`.

Comment (josephst) [reply .devops/nix/package.nix]: Man page for `xcrun` and `Xcode-select` [here](https://www.manpagez.com/man/1/xcode-select/)

**Comment by josephst** [comment]
Updated, based off [`macvim.nix` from Nixpkgs](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/editors/vim/macvim.nix), which also requires calling `xcrun` and cannot be built within a sandbox. It looks like setting `__noChroot = true` is the usual way to note that a package cannot be build within a sandboxed environment. 

Now, `/usr/bin/xcrun` is symlinked and provided as a `nativeBuildInput` when building on Darwin system. No other changes to $PATH are made.

File: .devops/nix/package.nix
Comment (SomeoneSerge) [inline .devops/nix/package.nix:166]: There's an undocumented feature [used in Nixpkgs](https://github.com/NixOS/nixpkgs/blob/2d5db19dffbb7fd86c8eba10a77deb8b8838ece3/pkgs/stdenv/generic/make-derivation.nix#L443-L448), called [`__impureHostDeps`](https://github.com/NixOS/nix/blob/5c8983b9e6bd47f1d921e21b717f936a02697344/src/libstore/build/local-derivation-goal.cc#L650-L652) which lets you demand specific paths from the host. That would be more fine-grained than disabling the sandbox.

That said, Nixpkgs seems to expose its own [`xcrun`](https://github.com/NixOS/nixpkgs/blob/2d5db19dffbb7fd86c8eba10a77deb8b8838ece3/pkgs/development/tools/xcbuild/wrapper.nix#L54-L106) as a passthru at `xcbuild.xcrun`: https://github.com/NixOS/nixpkgs/blob/2d5db19dffbb7fd86c8eba10a77deb8b8838ece3/pkgs/development/tools/xcbuild/wrapper.nix#L118. Are you saying it's broken? If it is we should report it

Comment (josephst) [reply .devops/nix/package.nix:166]: I don't think that `xcrun` shim will work here. Looking at source and testing on my system, [the `xcrun` shim from `xcbuild`](https://github.com/NixOS/nixpkgs/blob/2d5db19dffbb7fd86c8eba10a77deb8b8838ece3/pkgs/development/tools/xcbuild/wrapper.nix#L54-L106) expects whatever tool it is calling to be on `$PATH` so that the tool can be found with [`exec "$@"`](https://github.com/NixOS/nixpkgs/blob/2d5db19dffbb7fd86c8eba10a77deb8b8838ece3/pkgs/development/tools/xcbuild/wrapper.nix#L100C4-L100C13). This works for some binaries in `/usr/bin`, such as `cc` and `clang` (which are Apple-provided shims for the full binaries in `/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/`.

However, the `metal` compiler (and the directory it's in, `/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/metal`) are not added to $PATH because there may be multiple installs of Xcode on a system. `xcrun` handles running the `metal` binary from whichever Xcode installation is the default or provided via an environment variable. 

While investigating, it looks like the [`xcbuild` project from Facebook includes a more robust implementation of `xcrun`](https://github.com/facebookarchive/xcbuild/blob/dbaee552d2f13640773eb1ad3c79c0d2aca7229c/Libraries/xcsdk/Tools/xcrun.cpp) which can detect the install location of Xcode and run binaries from this location. From a brief glance at the code for that implementation, it looks through environment variables, the user's $HOME dir, and the standard install locations. 

That said, I don't think this ultimately solves the problem of needing an Xcode-specific binary to build llama.cpp on MacOS. Even with a shim for `xcrun`, the `metal` shader compiler from Xcode is still necessary. And if we're going to be calling Apple-provided binaries, I don't see any reason to avoid calling the normal `xcrun`, particularly as the [shell implementation](https://github.com/NixOS/nixpkgs/blob/2d5db19dffbb7fd86c8eba10a77deb8b8838ece3/pkgs/development/tools/xcbuild/wrapper.nix#L54-L106) only works for binaries already on $PATH, and the [`xcbuild` implementation from Facebook](https://github.com/facebookarchive/xcbuild/tree/master) is archived/ no longer maintained.

**Comment by josephst** [comment]
Just pushed a new version of this PR - still using Apple-provided `xcrun`, with `/usr/bin/xcrun` added to `__impureHostDeps` when creating the symlink. It's building and running properly on my system, including with sandboxing on (`nix build .# --option sandbox=true`)

File: .devops/nix/package.nix
Comment (SomeoneSerge) [reply .devops/nix/package.nix]: @josephst did you try the shim? Xcrun here is only used for the build, inside the sandbox, and it'll use the toolchain provisioned by nix

File: .devops/nix/package.nix
Comment (SomeoneSerge) [inline .devops/nix/package.nix:92]: NOTE: this will expose `/usr/bin/xcrun` to this particular `runCommand`, but not to the `mkDerivation` below (which is where you actually try to access xcrun)

File: .devops/nix/package.nix
Comment (josephst) [reply .devops/nix/package.nix]: Even when using the `xcrun` binary provided by Xcode/Apple, the rest of the build is using the toolchain provided by Nix and the final `./result/bin/llama` binary confirms that it is built with clang 16.0.6, provided by Nix toolchain (rather than clang 15.0.0, which is the system default installed with Xcode). 

The problem I'm having with the `xcrun` shim is that it cannot locate the `metal` shader compiler. `metal` is unfortunately proprietary so there's no nix equivalent (unlike ie `clang`; `xcrun clang` will look for `clang` on $PATH and find it as part of the nix toolchain)


To illustrate the problem:

With the "full" `xcrun` provided by MacOS/Xcode:
```
$ /usr/bin/xcrun -sdk macosx metal --version
Apple metal version 32023.155 (metalfe-32023.155)
Target: air64-apple-darwin23.4.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/metal/macos/bin
```

With the shim `xcrun`:
```
$ nix run nixpkgs#xcbuild.xcrun -- -sdk macosx metal --version
/nix/store/d0hnw5gy6mji9hqxcs564xx6zvba5m4s-xcrun/bin/xcrun: line 42: exec: metal: not found
```

Comment (SomeoneSerge) [reply .devops/nix/package.nix]: Ah, I just realized we haven't got an aarch64-darwin builder on github.

Could you please share the build logs? If your xcode/metal is from the host system and you had the sandbox enabled, how can the build access it?

Comment (josephst) [reply .devops/nix/package.nix]: Hmm looks like I didn't have the sandbox properly enabled earlier. Just set `sandbox = true` and getting some new build errors. Lots of output, so [added both options (`__impureHostDeps` vs `noChroot`)  to a Gist](https://gist.github.com/josephst/38d30ffdd2543209c1206ec1dc6a7c43). 

As I work on this more, I'm starting to think that `__noChroot = true` is the better option. Trying to add `/usr/bin/xcrun` to `__impureHostDeps` with `sandbox=true` generates several new errors about how `/usr/bin/xcrun` isn't one of [the allowedImpureHostPrefixes](https://github.com/NixOS/nix/blob/7c506432abab84d79744f4454aa20fe0a458e0fb/src/libstore/globals.cc#L83). 

With `sandbox = true` in my `nix.conf`, the only way to get this to build is to go back to using `__noChroot = effectiveStdenv.isDarwin && useMetalKit;`, then building with `nix build .# --rebuild --option sandbox relaxed`. Sandboxing on macOS with Nix defaults to `false` so builds should work on macOS with `sandbox = off` or `sandbox = "relaxed"`.

Comment (SomeoneSerge) [reply .devops/nix/package.nix]: I see. Thanks for the research! The last bit I fail to understand is what has changed that it's now required to use the `metal` compiler from `xcode`: the aarch64-darwin always linked MetalKit and it worked at the time of https://github.com/ggerganov/llama.cpp/pull/4605#issuecomment-1869692179

CC @philiptaron

Comment (josephst) [reply .devops/nix/package.nix]: Not exactly sure which commit changed it, but [looks like previously](https://github.com/ggerganov/llama.cpp/blob/58ba655af054715c0516ee270ad028ad9e74f357/CMakeLists.txt#L177) the `metal` compiler was only called if `LLAMA_METAL_SHADER_DEBUG` was true. 

Now, it's always called and `LLAMA_METAL_SHADER_DEBUG` affects which flags it's called with. Looks like the change was [about a week ago](https://github.com/ggerganov/llama.cpp/commit/381da2d9f0940d7009e3e918bed36338c8ff2fbb)

File: .devops/nix/package.nix
Comment (SomeoneSerge) [inline .devops/nix/package.nix:91]: Maybe rename to `xcrunHost`/`hostsXcrun`. I also think that `makeWrapper` was a better solution but it's ok if you keep this as is

File: .devops/nix/package.nix
Comment (SomeoneSerge) [inline .devops/nix/package.nix:166]: Add a comment with a permalink to the github discussion and the brief summary of the xcode/metal situation

Comment (SomeoneSerge) [reply .devops/nix/package.nix:166]: A link to the PR you found earlier, making `metal` mandatory, would also be nice

File: .devops/nix/package.nix
Comment (SomeoneSerge) [inline .devops/nix/package.nix:1]: Aside from the comments above, I'm only waiting for @philiptaron (he's got access to darwin) before merging. Thank you @josephst

**Comment by josephst** [comment]
Updated to incorporate latest changes. Thanks for the review, @SomeoneSerge!

**Comment by philiptaron** [comment]
I'll be able to run this either tonight (PDT) or tomorrow. Thanks for your patience.

**Comment by SomeoneSerge** [comment]
>  It looks like instead, the shader is converted to a string and embedded in the binary

This sounds like opengl/opencl too. Do you know if apple does any caching for kernels built on the fly?

I'd suggest we provide both options. E.g. make a `precompileMetalShaders ? false` flag for `callPackage`, disable chroot when `true`, otherwise set `DLLAMA_METAL_EMBED_LIBRARY`? I'd recommend making `false` the default because it's less likely to break. We can expose both as flake outputs.

**Comment by josephst** [comment]
@philiptaron actually, hold off on testing. It's building but as I run it I'm getting some unusual issues related to CMake. I think the final CMake install steps aren't copying `default.metallib` (that was just compiled) to the `$out/bin` directory. Need to investigate further.

**Comment by josephst** [comment]
Should be ready for testing now - will work on adding additional commit with option for embedding (and avoiding Xcode/xcrun dependency) with `precompileMetalShaders ? false` or something similar, but now builds on my system with `sandbox = relaxed` and runs well with Mistral 0.2 7B. 

This may also address #6020.

**Comment by swdunlop** [comment]
The following worked perfectly for me on MacOS Sonoma 14.3.1 / M2:

```
llama.cpp on  nix-darwin-xcrun via C v15.0.0-clang via △ via 🐍 v3.9.6 via 🐦 v5.9.2 via ↯ took 1m11s
$ nix run .#llama-server -- -m nomic-embed-text.gguf --embeddings
```

Thank you for chasing this the whole way through.

**Comment by josephst** [comment]
@philiptaron ready for review - ultimately broke this up into 3 commits (one for fixing `xcrun`, one for moving `default.metallib` into the `bin/` dir so that it can be loaded, one for enable/disable precompilation of the `.metal` files).

**Review by philiptaron — APPROVED** [review]

File: .devops/nix/package.nix
Comment (SomeoneSerge) [inline .devops/nix/package.nix:193]: `precompileMetalShaders` or we expose the broken symlink in `PATH`

**Comment by philiptaron** [comment]
I intend to merge once CI finishes with success.

**Review by SomeoneSerge — APPROVED** [review]

**Comment by philiptaron** [comment]
Sorry, I merged something else and now we have some conflicts.

File: .devops/nix/package.nix
Comment (SomeoneSerge) [inline .devops/nix/package.nix:239]: I was going to mention this before and now that there's a merge conflict I have a reason to: multiple `-DCMAKE_C_FLAGS` options aren't going to compose. I think we should move those to `NIX_CFLAGS_COMPILE`

**Comment by josephst** [comment]
Just rebased onto `master` so should be able to be merged

File: .devops/nix/package.nix
Comment (josephst) [reply .devops/nix/package.nix:240]: Are there any other uses of `CMAKE_C_FLAGS` currently? Also, could you elaborate on moving this option into `NIX_CFLAGS_COMPILE`? 

Right now, thinking of something like

```
let
  cmakeCFlags = lib.optionalString useMetalKit [ "-D__ARM_FEATURE_DOTPROD=1" ];
# ...
in {
# ...

cmakeFlags = # ...rest of section
  ++ lib.cmakeFeature "CMAKE_C_FLAGS" cmakeCFlags;
# ...
}
```

Which would allow other backends to configure `CMAKE_C_FLAGS` via appending to `cmakeCFlags`

Comment (SomeoneSerge) [reply .devops/nix/package.nix:240]: `NIX_CFLAGS_COMPILE` is an environment variable interpreted by nixpkgs' `cc-wrapper`. You can just pass it to `mkDerivation`: `NIX_CFLAGS_COMPILE = [ "-DA=1" "-DB=2" "-D__ARM_FEATURE_DOTPROD=1" ]`

> Are there any other uses of CMAKE_C_FLAGS currently?

I thought that was the merge conflict?

Comment (SomeoneSerge) [reply .devops/nix/package.nix:240]: I guess I was confused, it must've been just a change in formatting

Comment (josephst) [reply .devops/nix/package.nix:240]: ah, good to know about that variable. Merge conflict was from the following line (blas support) being deleted in master, likely a merge conflict line since I split the preceding line onto two lines

**Comment by philiptaron** [comment]
@SomeoneSerge you want to click the button this time? I think I don't need to wait for CI to pass given I saw it pass prior to the merge conflict.

**Pull request merged**
